File: test/test-junit5/src/test/java/ru/tinkoff/kora/test/extension/junit5/initializemode/ConstructorPerMethodTests.java
package ru.tinkoff.kora.test.extension.junit5.initializemode;

import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;
import ru.tinkoff.kora.test.extension.junit5.KoraAppTest;
import ru.tinkoff.kora.test.extension.junit5.TestComponent;
import ru.tinkoff.kora.test.extension.junit5.testdata.TestApplication;
import ru.tinkoff.kora.test.extension.junit5.testdata.TestComponent1;
import ru.tinkoff.kora.test.extension.junit5.testdata.TestComponent12;

import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNotSame;

@KoraAppTest(value = TestApplication.class)
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class ConstructorPerMethodTests {

    static volatile TestComponent1 prevComponent1;
    static volatile TestComponent12 prevComponent12;

    private final TestComponent1 component1;
    private final TestComponent12 component12;

    public ConstructorPerMethodTests(@TestComponent TestComponent1 component1,
                                     @TestComponent TestComponent12 component12) {
        this.component1 = component1;
        this.component12 = component12;
    }

    @Test
    @Order(1)
    void test1() {
        assertNotNull(component1);
        assertNotNull(component12);
        prevComponent1 = component1;
        prevComponent12 = component12;
    }

    @Test
    @Order(2)
    void test2() {
        assertNotNull(component1);
        assertNotNull(component12);
        assertNotSame(prevComponent1, component1);
        assertNotSame(prevComponent12, component12);
    }
}

File: test/test-junit5/src/test/kotlin/ru/tinkoff/kora/test/extension/junit5/kotlin/initializemode/ConstructorPerMethodTests.kt
package ru.tinkoff.kora.test.extension.junit5.kotlin.initializemode

import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNotSame
import ru.tinkoff.kora.test.extension.junit5.KoraAppTest
import ru.tinkoff.kora.test.extension.junit5.TestComponent
import ru.tinkoff.kora.test.extension.junit5.testdata.TestApplication
import ru.tinkoff.kora.test.extension.junit5.testdata.TestComponent1
import ru.tinkoff.kora.test.extension.junit5.testdata.TestComponent12
import kotlin.concurrent.Volatile

@KoraAppTest(value = TestApplication::class)
@TestInstance(TestInstance.Lifecycle.PER_METHOD)
@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
class ConstructorPerMethodTests(
    @TestComponent private val component1: TestComponent1,
    @TestComponent private val component12: TestComponent12
) {

    @Test
    @Order(1)
    fun test1() {
        assertNotNull(component1)
        assertNotNull(component12)
        prevComponent1 = component1
        prevComponent12 = component12
    }

    @Test
    @Order(2)
    fun test2() {
        assertNotNull(component1)
        assertNotNull(component12)
        assertNotSame(prevComponent1, component1)
        assertNotSame(prevComponent12, component12)
    }

    companion object {
        @Volatile
        var prevComponent1: TestComponent1? = null

        @Volatile
        var prevComponent12: TestComponent12? = null
    }
}
